### DISCUSSION TITLE
Permission Laravel error

### DISCUSSION THREAD

**Discussion by Pawelkal** [community/community#139747] [Programming Help] [answered]
### Body

Hi! I'm encountering an issue when running my Laravel project in Docker on a remote machine. Everything works fine locally, but on the server, I get the following error:

```php
file_put_contents(/var/www/storage/framework/views/...): Failed to open stream: Permission denied
```  

I'm using **Laravel 8**, the project runs on an **Ubuntu server**, and setup on the local machine are the same, but there's no issue there.

 
Could anyone help me figure out why this happens and how to fix it **in Docker file**?

### Guidelines

- [x] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by EugeneBachura** [comment, marked as answer]
### Problem: `Permission denied` when running Laravel in Docker on a remote machine

This error indicates that Laravel is unable to write to the `storage/framework/views` folder due to insufficient permissions on the server. While everything might work fine locally, the permissions on a remote machine, especially when using Docker, could be different.

### Solution:

#### 1. Check Permissions

Ensure that the `storage` folder and its subdirectories have the correct permissions. They need to be writable by the user under which Laravel is running.

Run the following commands:

```bash
sudo chmod -R 775 storage
sudo chmod -R 775 bootstrap/cache
sudo chown -R www-data:www-data storage
sudo chown -R www-data:www-data bootstrap/cache
```

#### 2. Fix on Docker Level

If the issue persists, you can address it within your Docker configuration.

##### a. Dockerfile Changes

In your `Dockerfile`, you can set the correct permissions while building the container. Add the following command:

```Dockerfile
RUN chown -R www-data:www-data /var/www/storage /var/www/bootstrap/cache
```

Or, if Laravel runs under a different user, adjust the permissions:

```Dockerfile
RUN chmod -R 775 /var/www/storage /var/www/bootstrap/cache
```

##### b. Docker Compose

If you're using Docker Compose, explicitly set the user that should run the application by adding the following lines to your `docker-compose.yml`:

```yaml
services:
  app:
    image: your_laravel_image
    user: "www-data"  # Specify the user here
    volumes:
      - ./storage:/var/www/storage
      - ./bootstrap/cache:/var/www/bootstrap/cache
```

##### c. Rebuild Docker Containers

After making the necessary changes, rebuild your Docker containers:

```bash
docker-compose down
docker-compose build
docker-compose up -d
```

### Important Notes:

- **Avoid overly permissive permissions** (e.g., 777) as this can lead to security vulnerabilities.
- **Ensure that the user inside the Docker container** has the required write permissions for the directories where Laravel stores temporary and cached files (e.g., `storage` and `bootstrap/cache`).

**Reply by Pawelkal** [reply]
Thank you, this instruction helped me
